File: 1. Implement a stack using array.py
"""
Task: Implement a stack using array

Time Complexity: push/pop/top O(1)
Space Complexity: O(n)
"""

class ArrayStack:
    def __init__(self, capacity: int):
        self._data = [None] * capacity
        self._cap = capacity
        self._top = -1

    def push(self, x: int) -> bool:
        if self._top + 1 >= self._cap:
            return False
        self._top += 1
        self._data[self._top] = x
        return True

    def pop(self):
        if self._top < 0:
            raise IndexError("pop from empty stack")
        val = self._data[self._top]
        self._top -= 1
        return val

    def top(self):
        if self._top < 0:
            raise IndexError("top from empty stack")
        return self._data[self._top]

    def is_empty(self) -> bool:
        return self._top < 0


if __name__ == "__main__":
    s = ArrayStack(10)
    s.push(1)
    s.push(2)
    print(s.pop())
    print(s.top())

File: 1. Implement a stack using array.cpp
/*
 Task: Implement a stack using array
*/

#include <bits/stdc++.h>
using namespace std;

class ArrayStack
{
   public:
    explicit ArrayStack(size_t capacity) : data(new int[capacity]), cap(capacity), top_idx(-1) {}

    ~ArrayStack() { delete[] data; }

    bool push(int x)
    {
        if (static_cast<size_t>(top_idx + 1) >= cap)
            return false;
        data[++top_idx] = x;
        return true;
    }

    bool pop(int& out)
    {
        if (top_idx < 0)
            return false;
        out = data[top_idx--];
        return true;
    }

    bool isEmpty() const { return top_idx < 0; }

    int top() const
    {
        if (top_idx < 0)
            throw runtime_error("empty");
        return data[top_idx];
    }

   private:
    int* data;
    size_t cap;
    int top_idx;
};

int main()
{
    ArrayStack s(10);
    s.push(1);
    s.push(2);
    int v;
    if (s.pop(v))
        cout << v << "\n";
    cout << s.top() << "\n";
    return 0;
}

// Time Complexity: push/pop/top O(1)
// Space Complexity: O(n)
